main.go

```diff
@@ -148,22 +151,83 @@ func getConnWithRetry(flags *flagStorage) (c *gcsx.Connection, err error) {
 	return
 }
 
-func createStorageHandle(flags *flagStorage) (storageHandle storage.StorageHandle, err error) {
-	tokenSrc, err := auth.GetTokenSource(context.Background(), flags.KeyFile, flags.TokenUrl, true)
+func createHttpClientObj(flags *flagStorage) (httpClient *http.Client, err error) {
+	var transport *http.Transport
+	// Using http1 makes the client more performant.
+	if flags.ClientProtocol == mountpkg.HTTP1 {
+		transport = &http.Transport{
+			MaxConnsPerHost:     flags.MaxConnsPerHost,
+			MaxIdleConnsPerHost: flags.MaxIdleConnsPerHost,
+			// This disables HTTP/2 in transport.
+			TLSNextProto: make(
+				map[string]func(string, *tls.Conn) http.RoundTripper,
+			),
+		}
+	} else {
+		// For http2, change in MaxConnsPerHost doesn't affect the performance.
+		transport = &http.Transport{
+			DisableKeepAlives: true,
+			MaxConnsPerHost:   flags.MaxConnsPerHost,
+			ForceAttemptHTTP2: true,
+		}
+	}
+
+	tokenSrc, err := createTokenSource(flags)
 	if err != nil {
-		err = fmt.Errorf("get token source: %w", err)
+		err = fmt.Errorf("while fetching tokenSource: %w", err)
 		return
 	}
 
+	// Custom http client for Go Client.
+	httpClient = &http.Client{
+		Transport: &oauth2.Transport{
+			Base:   transport,
+			Source: tokenSrc,
+		},
+		Timeout: flags.HttpClientTimeout,
+	}
+
+	// Setting UserAgent through RoundTripper middleware
+	httpClient.Transport = &storage.UserAgentRoundTripper{
+		Wrapped:   httpClient.Transport,
+		UserAgent: getUserAgent(flags.AppName),
+	}
+
+	return httpClient, err
+}
+
+func isProdEndpoint(endpoint *url.URL) bool {
+	return (endpoint == nil) || (endpoint.Hostname() == GcsProdHostName)
+}
+
+func createTokenSource(flags *flagStorage) (tokenSrc oauth2.TokenSource, err error) {
+	if isProdEndpoint(flags.Endpoint) {
```

Comment:
Why is this necessary? When would we expect to explicitly pass in the prod endpoint?

Reply:
The default value of Endpoint is (`https://storage.googleapis.com/storage/v1`). See - [here](https://github.com/GoogleCloudPlatform/gcsfuse/blob/bb315266bb684dfd3ea9bed54cf530b467b1c023/flags.go#L136)

I'll make default value as nil, then we can remove that.

Reply:
Sounds good, yes there is some other things that can go on under the hood with endpoints in the client (e.g. using MTLS) so it's better to not pass the endpoint unless you want a non-prod one.